src/borg/repository.py

```diff
@@ -1319,28 +1319,52 @@ def _close_fd(self, ts_fd):
         safe_fadvise(fd.fileno(), 0, 0, 'DONTNEED')
         fd.close()
 
+    def get_segment_dirs(self, start_index=None, end_index=None):
+        """Returns generator yeilding required segment dirs as `os.DirEntry` objects. Start and end are inclusive.
+        """
+        data_dir = os.path.join(self.path, 'data')
+        if start_index is None and end_index is None:
+            segment_dirs = (f for f in os.scandir(data_dir) if f.is_dir() and f.name.isdigit())
+        elif start_index is not None and end_index is None:
+            segment_dirs = (f for f in os.scandir(data_dir) if f.is_dir() and f.name.isdigit() and start_index <= int(f.name))
+        elif start_index is None and end_index is not None:
+            segment_dirs = (f for f in os.scandir(data_dir) if f.is_dir() and f.name.isdigit() and int(f.name) <= end_index)
+        elif start_index is not None and end_index is not None:
+            segment_dirs = (f for f in os.scandir(data_dir) if f.is_dir() and f.name.isdigit() and start_index <= int(f.name) <= end_index)
```

Comment:
guess this could be simplified a lot by not using None as the default values for start and end.

you could use start=0 and end=2**30 (the later is a bit arbitrary, but should still fit into a 64bit integer easily and be way more than we'll ever need).

Reply:
Yes. I was actually doing it for performance reasons. 

The comparison `start_index <= int(f.name) <= end_index)` is done on every directory, so for a large number of directories, it could add an *unnecessary* hit when a constrained list is not in fact required (like say for clear_empty_dirs function).

```
# Timing listing from generator with and without index bounds comparision

[tmp]$ python3 -m timeit -s "import os; data_dir = os.path.join('borg-100/', 'data'); start, end = 0, 2**30"  "list((f for f in os.scandir(data_dir) if f.is_dir() and f.name.isdigit()))" 
10000 loops, best of 5: 22 usec per loop

[tmp]$ python3 -m timeit -s "import os; data_dir = os.path.join('borg-100/', 'data'); start, end = 0, 2**30"  "list((f for f in os.scandir(data_dir) if f.is_dir() and f.name.isdigit() and start <= int(f.name) <= end))" 
10000 loops, best of 5: 31.9 usec per loop 

# Initial Conclusion: Adding the index bound comparision adds a 50% increase in timing for borg data dir with 100 segment directories.
```

The different cases were added, so we could return the most efficient generator for the specific use case. 

```

# For a 500GiB borg repo we'd have approximately ~ 1000 segment files. 
# So similarly testing for 1000 segment dirs

[tmp]$ python3 -m timeit -s "import os; data_dir = os.path.join('borg-1000/', 'data'); start, end = 0, 2**30"  "list((f for f in os.scandir(data_dir) if f.is_dir() and f.name.isdigit()))" 
2000 loops, best of 5: 195 usec per loop

[tmp]$ python3 -m timeit -s "import os; data_dir = os.path.join('borg-1000/', 'data'); start, end = 0, 2**30"  "list((f for f in os.scandir(data_dir) if f.is_dir() and f.name.isdigit() and start <= int(f.name) <= end))" 
1000 loops, best of 5: 303 usec per loop

Conclusion: An understandably greater time difference for larger repos.   
But in total only about 0.1 msec per full scan. 
```

In summary, around 0.1 millisecond saving per full scan for larger repos.  So maybe not enough to matter after all?
I'll shorten the code.

Reply:
guess time needed for the str to integer conversion and 2 integer comparisons is way less than the usual I/O overhead.